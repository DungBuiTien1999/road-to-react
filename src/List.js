import React, { useState } from 'react';
import _ from 'lodash';
import styles from './App.module.css';
import { ReactComponent as Check } from './check.svg';

export const Item = ({ item, onRemoveItem }) => {
  return (
    <div className={styles.item}>
      <span style={{ width: '40%' }}>
        <a href={item.url}>{item.title}</a>
      </span>
      <span style={{ width: '30%' }}>{item.author}</span>
      {/* <span style={{ width: '10%' }}>{item.points}</span> */}
      <span style={{ width: '10%' }}>{item.num_comments}</span>
      <span style={{ width: '10%' }}>{item.points}</span>
      <span style={{ width: '10%' }}>
        <button
          type="button"
          onClick={() => onRemoveItem(item)}
          className={`${styles.button} ${styles.buttonSmall}`}
        >
          <Check width="18px" height="18px" />
        </button>
      </span>
    </div>
  );
};

const SORTS = {
  NONE: (list) => list,
  TITLE: (list) => _.sortBy(list, 'title'),
  AUTHOR: (list) => _.sortBy(list, 'author'),
  COMMENTS: (list) => _.sortBy(list, 'num_comments').reverse(),
  POINTS: (list) => _.sortBy(list, 'points').reverse(),
};

const List = ({ stories, onRemoveItem }) => {
  const [sort, setSort] = useState('NONE');
  const [active, setActive] = useState('');

  const handleSort = (sortKey) => {
    setSort(sortKey);
  };

  const sortFunction = SORTS[sort];
  const sortListed = sortFunction(stories);

  return (
    <div>
      <div style={{ display: 'flex' }}>
        <span style={{ width: '40%' }}>
          <button type="button" onClick={() => handleSort('TITLE')}>
            Title
          </button>
        </span>
        <span style={{ width: '30%' }}>
          <button type="button" onClick={() => handleSort('AUTHOR')}>
            Author
          </button>
        </span>
        <span style={{ width: '10%' }}>
          <button type="button" onClick={() => handleSort('COMMENTS')}>
            Comments
          </button>
        </span>
        <span style={{ width: '10%' }}>
          <button type="button" onClick={() => handleSort('POINTS')}>
            Points
          </button>
        </span>
        <span style={{ width: '10%' }}>Actions</span>
      </div>
      {sortListed.map((item) => (
        <Item key={item.objectID} onRemoveItem={onRemoveItem} item={item} />
      ))}
    </div>
  );
};

export default List;
