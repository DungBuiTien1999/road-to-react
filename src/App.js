import React, {
  useState,
  useEffect,
  useRef,
  useReducer,
  useCallback,
} from 'react';
import axios from 'axios';
import styles from './App.module.css';
// import styled from "styled-components";
import List from './List';
import SearchForm from './SearchForm';

// const StyledContainer = styled.div`
//   height: 100vh;
//   padding: 2rem;

//   background: #83a4d4;
//   background: linear-gradient(to left, #b6fbff, #83a4d4);

//   color: #171212;
//   font-size: 1.6rem;
// `;

// const StyledHeadlinePrimary = styled.h1`
//   font-size: 8.4rem;
//   font-weight: 300;
//   letter-spacing: 0.2rem;
// `;

// const initialStories = [
//   {
//     title: "React",
//     url: "https://reactjs.org/",
//     author: "Jordan Walke",
//     num_comments: 3,
//     points: 4,
//     objectID: 0,
//   },
//   {
//     title: "Redux",
//     url: "https://redux.js.org/",
//     author: "Dan Abramov, Andrew Clark",
//     num_comments: 2,
//     points: 5,
//     objectID: 1,
//   },
// ];

const useSemiPersistentState = (key, initialState) => {
  const isMounted = useRef(false);
  const [value, setValue] = useState(localStorage.getItem(key) || initialState);

  useEffect(() => {
    if (!isMounted.current) {
      isMounted.current = true;
    } else {
      console.log('A');
      localStorage.setItem(key, value);
    }
  }, [value, key]);

  return [value, setValue];
};

// const getSumComments = (stories) => {
//   console.log("C");

//   return stories.data.reduce((result, value) => result + value.num_comments, 0);
// };

const storiesReducer = (state, action) => {
  switch (action.type) {
    case 'STORIES_FETCH_INIT':
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case 'STORIES_FETCH_SUCCESS':
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case 'STORIES_FETCH_FAILURE':
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case 'REMOVE_STORY':
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        ),
      };
    default:
      throw new Error();
  }
};

const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query=';

const App = () => {
  const [searchTerm, setSearchTerm] = useSemiPersistentState('search', 'React');
  const [url, setUrl] = useState(`${API_ENDPOINT}${searchTerm}`);
  const [stories, dispatchStories] = useReducer(storiesReducer, {
    data: [],
    isLoading: false,
    isError: false,
  });
  // const [isLoading, setIsLoading] = useState(false);
  // const [isError, setIsError] = useState(false);

  const handleSearchInput = (event) => {
    setSearchTerm(event.target.value);
  };

  const handleSearchSubmit = (e) => {
    setUrl(`${API_ENDPOINT}${searchTerm}`);

    e.preventDefault();
  };

  // const getAsyncStories = () =>
  //   new Promise((resolve, reject) => {
  //     setTimeout(() => resolve({ data: { stories: initialStories } }), 2000);
  //   });

  const handleFetchStories = useCallback(async () => {
    dispatchStories({ type: 'STORIES_FETCH_INIT' });
    try {
      const result = await axios.get(url);
      dispatchStories({
        type: 'STORIES_FETCH_SUCCESS',
        payload: result.data.hits,
      });
    } catch {
      dispatchStories({ type: 'STORIES_FETCH_FAILURE' });
    }
  }, [url]);

  useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  // const searchedStories = stories.data.filter((story) =>
  //   story.title.toLowerCase().includes(searchTerm.toLowerCase())
  // );

  const handleRemoveStory = (item) => {
    dispatchStories({
      type: 'REMOVE_STORY',
      payload: item,
    });
  };

  // const sumComments = useMemo(() => getSumComments(stories), [stories]);

  return (
    <div className={styles.container}>
      <h1 className={styles.headlinePrimary}>My Hacker Stories</h1>

      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />

      <hr />
      {stories.isError && <p>Sonething went wrong ...</p>}
      {stories.isLoading ? (
        <p>Loading ...</p>
      ) : (
        <List stories={stories.data} onRemoveItem={handleRemoveStory} />
      )}
    </div>
  );
};

export default App;
