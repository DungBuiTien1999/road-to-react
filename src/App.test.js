import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';
import { Item } from './List';
import List from './List';
import SearchForm from './SearchForm';
import axios from 'axios';

jest.mock('axios');

describe('Item', () => {
  const item = {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
  };
  const handleRemoveItem = jest.fn();

  let component;

  beforeEach(() => {
    component = renderer.create(
      <Item item={item} onRemoveItem={handleRemoveItem} />
    );
  });

  it('renders all properties', () => {
    const component = renderer.create(<Item item={item} />);

    expect(component.root.findByType('a').props.href).toEqual(
      'https://reactjs.org/'
    );
    // expect(component.root.findAllByType('span')[1].props.children).toEqual(
    //   'Jordan Walke'
    // ); // but this way isn't not thoughout when change the order of spans

    expect(
      component.root.findAllByProps({ children: 'Jordan Walke' }).length
    ).toEqual(1);
  });

  it('call onRemoveItem on button click', () => {
    component.root.findByType('button').props.onClick();

    expect(handleRemoveItem).toHaveBeenCalledTimes(1);
    expect(handleRemoveItem).toHaveBeenCalledWith(item);

    expect(component.root.findAllByType(Item).length).toEqual(1);
  });

  it('renders snapshot', () => {
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('List', () => {
  const list = [
    {
      title: 'React',
      url: 'https://reactjs.org/',
      author: 'Jordan Walke',
      num_comments: 3,
      points: 4,
      objectID: 0,
    },
    {
      title: 'Redux',
      url: 'https://redux.js.org/',
      author: 'Jordan Walke',
      num_comments: 2,
      points: 5,
      objectID: 1,
    },
  ];

  let component;

  beforeEach(() => {
    component = renderer.create(<List stories={list} />);
  });

  it('render two items', () => {
    expect(component.root.findAllByType(Item).length).toEqual(2);
  });

  it('renders snapshot', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('SearchForm', () => {
  const searchFormProps = {
    searchTerm: 'React',
    onSearchInput: jest.fn(),
    onSearchSubmit: jest.fn(),
  };

  let component;

  beforeEach(() => {
    component = renderer.create(<SearchForm {...searchFormProps} />);
  });

  it('renders the input field with its value', () => {
    const value = component.root.findByType('input').props.value;

    expect(value).toEqual('React');
  });

  it('changes the input field', () => {
    const pseodoEvent = { target: 'redux' };

    component.root.findByType('input').props.onChange(pseodoEvent);

    expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
    expect(searchFormProps.onSearchInput).toHaveBeenCalledWith(pseodoEvent);
  });

  it('submits the form', () => {
    const pseodo = {};

    component.root.findByType('form').props.onSubmit(pseodo);

    expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
    expect(searchFormProps.onSearchSubmit).toHaveBeenCalledWith(pseodo);
  });

  it('disables the button and prevent submit', () => {
    component.update(<SearchForm {...searchFormProps} searchTerm="" />);

    expect(component.root.findByType('button').props.disabled).toBeTruthy();
  });

  it('renders snapshot', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('App', () => {
  it('succeeds fetching data with a list', async () => {
    const list = [
      {
        title: 'React',
        url: 'https://reactjs.org/',
        author: 'Jordan Walke',
        num_comments: 3,
        points: 4,
        objectID: 0,
      },
      {
        title: 'Redux',
        url: 'https://redux.js.org/',
        author: 'Jordan Walke',
        num_comments: 2,
        points: 5,
        objectID: 1,
      },
    ];

    const promise = Promise.resolve({
      data: {
        hits: list,
      },
    });

    axios.get.mockImplementationOnce(() => promise);

    let component;

    await renderer.act(async () => {
      component = renderer.create(<App />);
    });

    expect(component.root.findByType(List).props.stories).toEqual(list);
  });

  it('fails fetching data with a list', async () => {
    const promise = Promise.reject();

    axios.get.mockImplementationOnce(() => promise);

    let component;

    await renderer.act(async () => {
      component = renderer.create(<App />);
    });

    expect(component.root.findByType('p').props.children).toEqual(
      'Sonething went wrong ...'
    );
  });

  it('renders snapshot', async () => {
    const list = [
      {
        title: 'React',
        url: 'https://reactjs.org/',
        author: 'Jordan Walke',
        num_comments: 3,
        points: 4,
        objectID: 0,
      },
      {
        title: 'Redux',
        url: 'https://redux.js.org/',
        author: 'Jordan Walke',
        num_comments: 2,
        points: 5,
        objectID: 1,
      },
    ];

    const promise = Promise.resolve({
      data: {
        hits: list,
      },
    });

    axios.get.mockImplementationOnce(() => promise);

    let component;

    await renderer.act(async () => {
      component = renderer.create(<App />);
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
