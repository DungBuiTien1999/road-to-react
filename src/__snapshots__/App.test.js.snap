// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders snapshot 1`] = `
<div
  className="container"
>
  <h1
    className="headlinePrimary"
  >
    My Hacker Stories
  </h1>
  <form
    className="searchForm"
    onSubmit={[Function]}
  >
    <div>
      <label
        className="label"
        htmlFor="search"
      >
        <strong>
          Search
        </strong>
        :
         
      </label>
       
      <input
        className="input"
        id="search"
        onChange={[Function]}
        type="text"
        value="React"
      />
    </div>
    <button
      className="button buttonLarge"
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </form>
  <hr />
  <div
    className="item"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href="https://reactjs.org/"
      >
        React
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      Jordan Walke
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      3
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      4
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button buttonSmall"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="18px"
          width="18px"
        >
          check.svg
        </svg>
      </button>
    </span>
  </div>
  <div
    className="item"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href="https://redux.js.org/"
      >
        Redux
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      Jordan Walke
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      5
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button buttonSmall"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="18px"
          width="18px"
        >
          check.svg
        </svg>
      </button>
    </span>
  </div>
</div>
`;

exports[`Item renders snapshot 1`] = `
<div
  className="item"
>
  <span
    style={
      Object {
        "width": "40%",
      }
    }
  >
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span
    style={
      Object {
        "width": "30%",
      }
    }
  >
    Jordan Walke
  </span>
  <span
    style={
      Object {
        "width": "10%",
      }
    }
  >
    3
  </span>
  <span
    style={
      Object {
        "width": "10%",
      }
    }
  >
    4
  </span>
  <span
    style={
      Object {
        "width": "10%",
      }
    }
  >
    <button
      className="button buttonSmall"
      onClick={[Function]}
      type="button"
    >
      <svg
        height="18px"
        width="18px"
      >
        check.svg
      </svg>
    </button>
  </span>
</div>
`;

exports[`List renders snapshot 1`] = `
Array [
  <div
    className="item"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href="https://reactjs.org/"
      >
        React
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      Jordan Walke
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      3
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      4
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button buttonSmall"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="18px"
          width="18px"
        >
          check.svg
        </svg>
      </button>
    </span>
  </div>,
  <div
    className="item"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href="https://redux.js.org/"
      >
        Redux
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      Jordan Walke
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      5
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button buttonSmall"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="18px"
          width="18px"
        >
          check.svg
        </svg>
      </button>
    </span>
  </div>,
]
`;

exports[`SearchForm renders snapshot 1`] = `
<form
  className="searchForm"
  onSubmit={[MockFunction]}
>
  <div>
    <label
      className="label"
      htmlFor="search"
    >
      <strong>
        Search
      </strong>
      :
       
    </label>
     
    <input
      className="input"
      id="search"
      onChange={[MockFunction]}
      type="text"
      value="React"
    />
  </div>
  <button
    className="button buttonLarge"
    disabled={false}
    type="submit"
  >
    Submit
  </button>
</form>
`;
